x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1933,4 +1954,154 @@ private static void typeResolutions(
+        }
+
+        private void typeResolutions(
+            InvalidMappedField imf,
+            AggregateFunction aggregateFunction,
+            FieldAttribute fa,
+            Expression expression,
+            AggregateMetricDoubleBlockBuilder.Metric metric,
+            HashMap<ResolveUnionTypes.TypeResolutionKey, Expression> typeResolutions
+        ) {
+            for (DataType type : imf.types()) {
+                // Effectively the contents of ResolveUnionTypes::typeSpecificConvert(...)
+                // except convertFunction is not necessarily a ConvertFunction (as in the case of Sum's FromAggregateMetricDouble)
+                // and we do not substitute surrogates because Count does have a surrogate in the case of aggregate metric double
+                ResolveUnionTypes.TypeResolutionKey key = new ResolveUnionTypes.TypeResolutionKey(fa.name(), type);
+                EsField field = new EsField(imf.getName(), type, imf.getProperties(), imf.isAggregatable());
+                FieldAttribute originalFieldAttr = (FieldAttribute) aggregateFunction.field();
+                FieldAttribute resolved = new FieldAttribute(
+                    fa.source(),
+                    originalFieldAttr.parentName(),
+                    originalFieldAttr.name(),
+                    field,
+                    originalFieldAttr.nullable(),
+                    originalFieldAttr.id(),
+                    true
+                );
+
+                Expression convertExpression;
+                // Counting on aggregate metric double has unique behavior in that we cannot just provide the number of
+                // documents, instead we have to look inside the aggregate metric double's count field and sum those together.
+                // Grabbing the count value with FromAggregateMetricDouble the same way we do with min/max/sum would result in
+                // a single Int field, and incorrectly be treated as 1 document (instead of however many originally went into
+                // the aggregate metric double).
+                if (metric == AggregateMetricDoubleBlockBuilder.Metric.COUNT) {
+                    convertExpression = new ToAggregateMetricDouble(fa.source(), resolved);
+                } else if (type == AGGREGATE_METRIC_DOUBLE) {
+                    convertExpression = FromAggregateMetricDouble.withMetric(fa.source(), resolved, metric);
+                } else {
+                    convertExpression = new ToDouble(fa.source(), resolved);
+                }
+                Expression e = expression.replaceChildren(List.of(convertExpression, expression.children().get(1)));
+                typeResolutions.put(key, e.children().getFirst());
+            }
+        }
+
+        private Expression createIfDoesNotAlreadyExist(
+            FieldAttribute fa,
+            MultiTypeEsField resolvedField,
+            List<FieldAttribute> unionFieldAttributes,
+            AggregateMetricDoubleBlockBuilder.Metric metric
```

Comment:
Can we modify `createIfDoesNotAlreadyExist` instead of introducing this new method?